src/Tasks/Microsoft.NET.Build.Tasks/CreateAppHost.cs

```diff
@@ -28,34 +35,63 @@ public class CreateAppHost : TaskBase
 
         public bool WindowsGraphicalUserInterface { get; set; }
 
+        public int Retries { get; set; }
+
+        public int RetryDelayMilliseconds { get; set; } = DefaultRetryDelayMilliseconds;
+
         protected override void ExecuteCore()
         {
             try
             {
-                if (ResourceUpdater.IsSupportedOS())
+                var isGUI = WindowsGraphicalUserInterface;
+                var resourcesAssembly = IntermediateAssembly;
+
+                if (!ResourceUpdater.IsSupportedOS())
                 {
-                    HostWriter.CreateAppHost(appHostSourceFilePath: AppHostSourcePath,
-                                             appHostDestinationFilePath: AppHostDestinationPath,
-                                             appBinaryFilePath: AppBinaryName,
-                                             windowsGraphicalUserInterface: WindowsGraphicalUserInterface,
-                                             assemblyToCopyResorcesFrom: IntermediateAssembly);
+                    if (isGUI)
+                    {
+                        Log.LogWarning(Strings.AppHostCustomizationRequiresWindowsHostWarning);
+                    }
+
+                    isGUI = false;
+                    resourcesAssembly = null;
                 }
-                else
+
+                int attempts = 0;
+                
+                do
                 {
-                    // by passing null to assemblyToCopyResorcesFrom, it will skip copying resorces,
-                    // which is only supported on Windows
-                    if (WindowsGraphicalUserInterface)
+                    try
                     {
-                        Log.LogWarning(Strings.AppHostCustomizationRequiresWindowsHostWarning);
+                        HostWriter.CreateAppHost(appHostSourceFilePath: AppHostSourcePath,
+                                                appHostDestinationFilePath: AppHostDestinationPath,
+                                                appBinaryFilePath: AppBinaryName,
+                                                windowsGraphicalUserInterface: isGUI,
+                                                assemblyToCopyResorcesFrom: resourcesAssembly);
+                        return;
                     }
+                    catch (Exception ex) when (ex is IOException ||
+                                               ex is UnauthorizedAccessException)
+                                               //ex is ResourceUpdater.HResultException)
```

Comment:
This exception is currently private in core-setup.  We would need to update the `ResourceUpdater` to be able to catch it more specifically than `Exception`.

Reply:
Maybe `ex.GetType().Name == “HResultException”` for now against a bug on core-setup or were you going to PR the fix to core-setup and wait for it to flow. This is the case we’re hitting right? 

Also my suggestion would be for core-setup to make the public exception contract more abstract like ResourceUpdateFailedException or something. They could still include HResult in Exception.HResult property and message. This would allow the API to become x-plat without needing changes here.

Reply:
I can live with `ex.GetType().Name` check for now and file a core-setup bug to make the exception public (and better).

Reply:
Files as https://github.com/dotnet/core-setup/issues/7516.